src/replacer.rs

```diff
@@ -0,0 +1,334 @@
+    Some(Replacement {
+        fragments,
+        input,
+        output,
+    })
+}
+
+/// A replacement contains the of fragments, the input string and the output string
+#[derive(Debug)]
+pub struct Replacement<'a> {
+    fragments: Fragments,
+    input: &'a str,
+    output: String,
+}
+
+impl<'a> Replacement<'a> {
+    pub fn output(&self) -> &str {
+        &self.output
+    }
+
+    /// Print the replacement as two lines (red then green)
+    pub fn print_self(&self, prefix: &str) {
+        let red_underline = { |x: &str| x.red().underline() };
+        let input_fragments: Vec<_> = self.fragments.into_iter().map(|x| &x.0).collect();
+        Self::print_fragments(prefix, red_underline, self.input, &input_fragments);
+
+        let green_underline = { |x: &str| x.green().underline() };
+        let output_fragments: Vec<_> = self.fragments.into_iter().map(|x| &x.1).collect();
+        Self::print_fragments(prefix, green_underline, &self.output, &output_fragments);
+    }
+
+    fn print_fragments<C>(prefix: &str, color: C, line: &str, fragments: &[&Fragment])
+    where
+        C: Fn(&str) -> ColoredString,
+    {
+        print!("{}", prefix);
+        let mut current_index = 0;
+        for (i, fragment) in fragments.iter().enumerate() {
+            let Fragment { index, text } = fragment;
+            // Whitespace between prefix and the first fragment does not matter
+            if i == 0 {
+                print!("{}", &line[current_index..*index].trim_start());
+            } else {
+                print!("{}", &line[current_index..*index]);
+            }
+            print!("{}", color(text));
+            current_index = index + text.len();
+        }
+        println!("{}", &line[current_index..]);
+    }
+}
+
+#[derive(Debug)]
+// A list of input_fragment, output_fragent
+struct Fragments(Vec<(Fragment, Fragment)>);
+
+impl Fragments {
+    fn new() -> Self {
+        Self(vec![])
+    }
+
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    fn add(&mut self, input: (usize, &str), output: (usize, &str)) {
+        let input_fragment = Fragment {
+            index: input.0,
+            text: input.1.to_string(),
+        };
+        let output_fragent = Fragment {
+            index: output.0,
+            text: output.1.to_string(),
+        };
+        self.0.push((input_fragment, output_fragent));
+    }
+}
+
+impl<'a> IntoIterator for &'a Fragments {
+    type Item = &'a (Fragment, Fragment);
+
+    type IntoIter = std::slice::Iter<'a, (Fragment, Fragment)>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter()
+    }
+}
+
+/// Represent a fragment of text, similar to the data structure returned
+/// by String::match_indices
+#[derive(Debug)]
+struct Fragment {
+    index: usize,
+    text: String,
+}
+
+trait Replacer {
+    // return position of the match, input_text, output_text, or None
+    fn replace(&self, buff: &str) -> Option<(usize, String, String)>;
```

Comment:
Fun fact: This could also use &str and Cow to avoid allocations!

`fn replace<'a>(&'a self, buff: &'a str) -> Option<(usize, &str, Cow<str>)>;`

And then the implementation practically doesn't change (except a .into here and there), for ex:

```rust
impl<'a> Replacer for RegexReplacer<'a> {
    fn replace<'b>(&'b self, buff: &'b str) -> Option<(usize, &str, Cow<str>)> {
        let regex_match = self.regex.find(buff)?;
        let pos = regex_match.start();
        let input_text = regex_match.as_str();
        let output_text = self.regex.replacen(input_text, 1, self.replacement);
        Some((pos, input_text, output_text))
    }
}
```

Reply:
Yeah but the optimization makes less sense here. If there's no match, we'll just return None. If we return Some(), then we *know* the string has changed, unless you did something silly like replace the regex `old` with `$1`.

Reply:
Oh yeah, it's just that your `SubstringReplacer` and `SubvertReplacer` already have a `&'a str` for the replacement, so for example SubstringReplacer returns:
 `Some((pos, self.pattern.to_string(), self.replacement.to_string()))`

But since it has a &str internally, it could skip those two copies, since the replacement is always the same (it comes from the command line).  
Although I agree it doesn't make a lot of sense here, it's way beyond premature optimization, I'm just seeing places for `Cow` everywhere even if it's not necessary =)